### PR TITLE
firefoxPackages.tor-browser*, tor-browser-bundle: remove

### PR DESCRIPTION
**Pull request by flokli**
These are all based on firefox versions with known vulnerabilities
exploited in the wild.

We seriously shouldn't ship this in nixpkgs, especially not for
sensitive applications as the Tor Browser.

`tor-browser-bundle` is just a wrapper around
`firefoxPackages.tor-browser`, so let's remove it too.

`tor-browser-bundle-bin` is the much safer bet, which is individually
downloaded from `dist.torproject.org` and just `patchelf`-ed locally to
work on NixOS.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by alyssais — APPROVED** [review]
I do think it would be nice to build TBB ourselves at some point, especially since it’s reproducible, but it’s clearly not something we’re up to right now, so I agree with removing.

**Comment by flokli** [comment]
It's currently causing more harm than good.

File: pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix [hunk 1]
```diff
@@ -394,7 +391,28 @@ stdenv.mkDerivation rec {
...
+
+      Note that `tor-browser-bundle` package is not only built from source, but also bundles Tor
+      Browser differently from the official `tor-browser-bundle-bin` implementation. The official
+      Tor Browser is not a normal UNIX program and is heavily patched for its use in the Tor Browser
+      Bundle (which `tor-browser-bundle-bin` package then has to work around for the read-only
+      /nix/store). Meanwhile, `firefoxPackages.tor-browser` reverts all those patches, allowing
+      `firefoxPackages.tor-browser` to be used independently of the bundle, and then implements what
+      `tor-browser-bundle` needs for the bundling using a much simpler patch. See the
+      longDescription and expression of the `firefoxPackages.tor-browser` package for more info.
+    '';
```
Comment (alyssais) [inline pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix]: This is all about packages that we’re removing, isn’t it? Shouldn’t we drop this?

Comment (flokli) [reply pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix]: Oh, right - I was just blindly moving the description from the non-bin variant into here, because it was previously imported. Fix incoming.

File: pkgs/top-level/aliases.nix [hunk 2]
```diff
@@ -532,4 +532,9 @@ mapAliases ({
 
   # added 2019-10-28
   gnatsd = nats-server;
+
+  # added 2020-01-10
+  tor-browser-bundle = throw "use tor-browser-bundle-bin instead.";
+  # added 2020-01-10
+  tor-browser-unwrapped = throw "use tor-browser-bundle-bin instead.";
```
Comment (grahamc) [inline pkgs/top-level/aliases.nix]: What if we set it up to issue a warning of some sort, and then automatically did the right thing? Also:

```suggestion
  tor-browser-unwrapped = throw "tor-browser-unwrapped was removed because it was out of date and inadequately maintained. Please use tor-browser-bundle-bin instead. See #77452.";
```

File: pkgs/applications/networking/browsers/firefox/packages.nix [hunk 3]
```diff
@@ -196,94 +196,8 @@ in {
...
-
-    patches = [
-      missing-documentation-patch
-    ];
-  };
-
-  tor-browser = tor-browser-8-5;
+  tor-browser-7-5 = throw "use pkgs.tor-browser-bundle-bin instead.";
+  tor-browser-8-5 = throw "use pkgs.tor-browser-bundle-bin instead.";
+  tor-browser = throw "use pkgs.tor-browser-bundle-bin instead.";
```
Comment (grahamc) [inline pkgs/applications/networking/browsers/firefox/packages.nix]: Same type of feedback from https://github.com/NixOS/nixpkgs/pull/77452/files#r365346928 applies here :)

File: pkgs/top-level/aliases.nix (hunk 2 above)
Comment (flokli) [reply pkgs/top-level/aliases.nix]: I want to avoid automatically doing "the right thing here", but having people explicitly making the decision. I will update the message however.

**Review by grahamc — APPROVED** [review]

**Review by jonringer — APPROVED** [review]

**Comment by vcunat** [comment]
@oxij might be interested, even if it's merged already.

**Comment by oxij** [comment]
Yeah, mentioning me and @joachifm before merging would have been nice. Also, for vulnerable software there's a `knownVulnerabilities` mechanism, so I'd say :-1: to removing this completely anyway.

So, tl;dr: I'd say let's revert and mark those as vulnerable instead for now, while we work out the solution to the problems discussed below.

The main problem there is that tor-browser 9.0 became kinda broken after they integrated tor-button and tor-launcher into the browser itself. Firstly, tor-launcher is not a `.gitsubmodule`, thus the fetching of sources is kinda broken if you want to build it with the launcher. Secondly, the usual unbundling patches from SLNOS do work and the newer versions of it do build, but building with `--disable-tor-launcher` produces a broken tor-browser (about:settings crashes). Which, with the problems outlined below, resulted in the SLNOS maintainer of tor-browser patches throw a bit of a fit and resign. Which gives us the current state.

I've been fiddling with the code there to make `--disable-tor-launcher` work with some ugly patching around, and I even have a version that seems to work now (needs a couple more days of testing, though), but there are several problems.

In the first place, the idea of `firefoxPackages.tor-browser` was to make an app conforming to the UNIX-way, and the upstream tor-browser is now an non-UNIX agglomeration of many things (e.g., it wants to start tor daemon by itself now, thus throwing away the usual separate-user isolation and etc, which, btw, uncomfortably reminds me of libpulse of PulseAudio, ugh). Thus, even if we are okay with carrying a rather ugly patch to make `firefoxPackages.tor-browser` with `--disable-tor-launcher` actually work (which, I'd say, is probably okay), and, possibly, another patch to tame the integrated tor-button so that it would not throw a fit when working without direct access to tor daemon (which does not seem to be a problem now, but that needs testing), we would probably still need a separate version for our `tor-browser-bundle`, which would now need to be reworked to integrate with integrated tor-launcher (which I did not manage to successfully build yet).

So, in short, I can make fresh versions of `firefoxPackages.tor-browser` work in short order. But I can't make `tor-browser-bundle` work in short order, and I don't really want to spend time trying as `firefoxPackages.tor-browser` works for me.

/cc @joachifm on the latter point, I guess

**Comment by alyssais** [comment]
Are there any users of firefoxPackages.tor-browser other than yourself?

**Comment by oxij** [comment]
> Are there any users of firefoxPackages.tor-browser other than yourself?

Outside of SLNOS, I assume @joachifm is one such user, but other than that I don't think we have means to answer that question unless each user explicitly confesses. Unless Hydra now collects some statistics I'm not aware of.

In general, IMHO, firefoxPackages.tor-browser + uMatrix is the only sane web browser setup there currently is (reasonably secure, can reasonably resist fingerprinting, has proper origin isolation, can disable anything you want by default, including the newly Turing-complete CSS3 that can send results of computations to arbitrary servers via media queries, thus making it another arbitrary code interpreter in the browser, does not send stuff to any corporate servers, half the web MITMed by CloudFlare does not throw infinite Google CAPTCHAs at you, etc). I actually use it as my default browser, even when not using tor, because, IMHO, it's what firefox should have been.

So, if there are no other users of it in NixOS, which I doubt, given it worked perfectly okay and was updated within a week of upstream release for the previous 2.5 years, I can only shake my head in pity and regret at lost opportunities.

**Comment by flokli** [comment]
@oxij I don't think there's any pressure to revert immediately. For 19.09, derivations are marked as insecure, and users are directed to `tor-browser-bundle-bin`.

`tor-browser-bundle` currently is vulnerable to at least `CVE-2019-17026`, and based on an unsupported firefox version. I wouldn't advertise it as "the only sane web browser setup there currently is" (because of the vulnerabilities), advise against using it for online browsing, and would feel much more comfortable if people would use `tor-browser-bundle-bin`.

I appreciate the effort you and the SLNOS community took into packaging tor-browser from source, but at least for nixpkgs, maintaining all the different flavours in the firefox expression has become a huge maintenance burden, especially if we're talking about keeping extra switches for versions already out of support of firefox by itself.
If people really need to use these old unsupported versions for some obscure reasons (like needing support for some never-updated plugins) we should defer to an older nixpkgs channel.

As far as Tor Browser is concerned, I'd also like if we'd be able to provide a self-built Tor Browser. However, apart from the fact this needs to be done with a lot of caution, it should be based on a currently supported firefox version, maybe simply a set of patches on top. As you said by yourself, given the current SLNOS tor-browser maintainer kinda resigned, I don't feel like this should be introduced before these issues have been adressed.

**Comment by oxij** [comment]
> sanity

You assume old `tor-browser`, my reasoning assumes `tor-browser` version 9.0.4, like in #77507.

> maintaining all the different flavours in the firefox expression has become a huge maintenance burden

IceCat takes literally two conditionals in `common.nix`, TorBrowser takes a bunch, but almost all of them are localized in `configureFlags` and `buildInputs`. Most of the burden there comes from different versions of the `firefox` itself, IceCat and TorBrowser burdens are uber-tiny compared to what they provide, IMHO.

Anyway, see #77507.

**Comment by flokli** [comment]
Yes, this is not only about tor-browser, but about keeping code for unsupported firefox versions in the tree.

**Comment by ehmry** [comment]
Note that unless we build ourselves there is no tor-browser for ARM.
